### PR TITLE
1278

### PR DESCRIPTION
**Pull request by Sild**
part of 1278 tests ( cover the metric & entity cases )
implements  `MetricMethod`, `EntityMethod`, `SeriesMethod` common methods ( like `insertT`, `getT`, `insertTCheck` etc... )
in addition add comments for tests without task number

### CODE REVIEW

File: src/test/java/com/axibase/tsd/api/method/metric/MetricCreateOrReplaceTest.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    public void testCreateOrReplace() throws Exception {
+        final Metric metric = new Metric("m-create-or-replace");
+        metric.setDataType(DataType.DECIMAL);
+
+        Response response = createOrReplaceMetric(metric.getName(), metric);
+        assertEquals(OK.getStatusCode(), response.getStatus());
+        assertTrue(metricExist(metric));
+    }
+
+    @Test //#1278
```
Comment (Sild) [inline src/test/java/com/axibase/tsd/api/method/metric/MetricCreateOrReplaceTest.java]: fail

File: src/test/java/com/axibase/tsd/api/Util.java [hunk 2]
```diff
@@ -16,6 +16,8 @@
 public class Util {
 
     private static final Long MILLIS_IN_DAY = 1000 * 60 * 60 * 24L;
+    public static final Long CHECK_INTERVAL = 500L;
+    public static final Long DEFAULT_CHECK_TIMEOUT = 2000L;
```
Comment (rmakulov) [inline src/test/java/com/axibase/tsd/api/Util.java]: why have you chosen such constants?

Comment (Sild) [inline src/test/java/com/axibase/tsd/api/Util.java]: empirical observations

File: src/test/java/com/axibase/tsd/api/method/entity/EntityCreateOrReplaceTest.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ */
+public class EntityCreateOrReplaceTest extends EntityMethod {
+
+    @Test //1278
+    public void testURLCreateOrReplaceWhitespace() throws Exception {
+        Entity entity = new Entity("createentity 1");
+
+        assertEquals(BAD_REQUEST.getStatusCode(), createOrReplaceEntity(entity).getStatus());
+    }
+
```
Comment (rmakulov) [inline src/test/java/com/axibase/tsd/api/method/entity/EntityCreateOrReplaceTest.java]: Could you please describe the test goal?

Comment (Sild) [inline src/test/java/com/axibase/tsd/api/method/entity/EntityCreateOrReplaceTest.java]: Trying to create entity which contain the whitespace in `metricName`. Make sure server deny this action

Comment (rmakulov) [inline src/test/java/com/axibase/tsd/api/method/entity/EntityCreateOrReplaceTest.java]: Ok, thank you. What does it mean URL in test name?

Comment (Sild) [inline src/test/java/com/axibase/tsd/api/method/entity/EntityCreateOrReplaceTest.java]: All tests are grouped by the following pattern: they use model fields in path (as part of url)

File: src/test/java/com/axibase/tsd/api/method/metric/MetricDeleteTest.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    @Test //#1278
+    public void testURLEncodeNameCyrillic() throws Exception {
+        final Metric metric = new Metric("deleteйёmetric-3");
+        createOrReplaceMetricCheck(metric);
+
+        assertEquals(OK.getStatusCode(), deleteMetric(metric.getName()).getStatus());
+        assertFalse(metricExist(metric));
+    }
+
+    @Test
```
Comment (rmakulov) [inline src/test/java/com/axibase/tsd/api/method/metric/MetricDeleteTest.java]: Test without ticket?

Comment (Sild) [inline src/test/java/com/axibase/tsd/api/method/metric/MetricDeleteTest.java]: Thanks, I'll add the corresponding comment.
Yes

File: src/test/java/com/axibase/tsd/api/method/series/SeriesQueryTest.java [hunk 5]
```diff
@@ -83,7 +88,7 @@ public void testLocalTimeUnsupported() throws Exception {
 
         seriesQuery.setStartDate("2016-07-01 14:23:20");
 
-        Response response = executeQueryReturnResponse(seriesQuery);
+        Response response = getSeries(seriesQuery);
```
Comment (rmakulov) [inline src/test/java/com/axibase/tsd/api/method/series/SeriesQueryTest.java]: Why getSeries returns Response?)

**Comment by rmakulov** [comment]
end review

File: src/test/java/com/axibase/tsd/api/Util.java (hunk 2 above)
Comment (rmakulov) [inline src/test/java/com/axibase/tsd/api/Util.java]: What use-cases for them?

Comment (Sild) [inline src/test/java/com/axibase/tsd/api/Util.java]: https://github.com/axibase/atsd-api-test/pull/42/commits/142071c6f836c8f417f9e20d854035fc6f7164df#diff-b367f829f39b804d4aaf3378f760e4fcR41

https://github.com/axibase/atsd-api-test/pull/42/commits/142071c6f836c8f417f9e20d854035fc6f7164df#diff-b367f829f39b804d4aaf3378f760e4fcR52

Wait for data to be handled by ATSD

File: src/test/java/com/axibase/tsd/api/method/series/SeriesQueryTest.java (hunk 5 above)
Comment (Sild) [inline src/test/java/com/axibase/tsd/api/method/series/SeriesQueryTest.java]: It's the default behavior of base queries, please look at 
https://github.com/axibase/atsd-api-test/blob/master/src/test/java/com/axibase/tsd/api/method/property/PropertyMethod.java#L30

https://github.com/axibase/atsd-api-test/blob/master/src/test/java/com/axibase/tsd/api/method/property/PropertyMethod.java#L39

This pattern allow to get flexible base methods for further processing

File: src/test/java/com/axibase/tsd/api/Util.java (hunk 2 above)
Comment (rmakulov) [inline src/test/java/com/axibase/tsd/api/Util.java]: Lets call them in appropriate way.

File: src/test/java/com/axibase/tsd/api/method/entity/EntityMethod.java [hunk 6]
```diff
@@ -46,13 +44,57 @@ public static Response getEntity(String entityName) {
...
+        WebTarget target = httpApiResource.path(METHOD_ENTITY_METRICS).resolveTemplate("entity", entityName);
+        for (Map.Entry<String, String> entry : parameters.entrySet()) {
+            target.queryParam(entry.getKey(), entry.getValue());
+        }
+        Response response = target.request().get();
+        response.bufferEntity();
+        return response;
+    }
+
+    public static Response getMetrics(String entityName) {
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/entity/EntityMethod.java]: May be to  replace it with `metrcs()`, `groups()`, `propertyTypes()`?

File: src/test/java/com/axibase/tsd/api/method/entity/EntityUpdateTest.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        createOrReplaceEntityCheck(entity);
+
+        Map<String, String> newTags = new HashMap<>();
+        newTags.put("t2", "tv2");
+        checkUrlencodedPathHandledSuccessfullyOnUpdate(entity, newTags);
+    }
+
+    private void checkUrlencodedPathHandledSuccessfullyOnUpdate(final Entity entity, Map newTags) throws Exception {
+        entity.setTags(newTags);
+        assertEquals(OK.getStatusCode(), updateEntity(entity).getStatus());
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/entity/EntityUpdateTest.java]: May be to add some assert message and raise special exception with explaining  message?

File: src/test/java/com/axibase/tsd/api/method/entity/EntityUpdateTest.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        Entity entity = new Entity("update_йёentity3");
+        entity.addTag("t1", "tv1");
+        createOrReplaceEntityCheck(entity);
+
+        Map<String, String> newTags = new HashMap<>();
+        newTags.put("t2", "tv2");
+        checkUrlencodedPathHandledSuccessfullyOnUpdate(entity, newTags);
+    }
+
+    private void checkUrlencodedPathHandledSuccessfullyOnUpdate(final Entity entity, Map newTags) throws Exception {
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/entity/EntityUpdateTest.java]: why check instead assert?

**Comment by unrealwork** [comment]
the review ended

File: src/test/java/com/axibase/tsd/api/method/entity/EntityUpdateTest.java (hunk 7 above)
Comment (Sild) [inline src/test/java/com/axibase/tsd/api/method/entity/EntityUpdateTest.java]: will be fixed to `assertEquals("Fail to execute updateEntity request", OK.getStatusCode(), updateEntity(entity).getStatus());`

**Comment by unrealwork** [comment]
ok

**Comment by rmakulov** [comment]
Ok

**Comment by KutuzovGregory** [comment]
ok

File: src/test/java/com/axibase/tsd/api/method/property/PropertyQueryOffsetTest.java [hunk 9]
```diff
@@ -33,6 +33,32 @@ public static void prepareProperty() throws IOException {
...
+            property.addKey(key[i], key[i + 1]);
+        }
+        property.addTag("defaultname", "defaultval");
+        property.setDate(date);
+
+        return property;
+    }
+
+    private static Response executeOffsetQuery(int offset) {
+        Map<String, Object> queryObj = new HashMap<>();
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/property/PropertyQueryOffsetTest.java]: May be to replace it with Collections intialize:

```java
Map<String, Object> queryObj = Collections.unmodifiableMap(new HashMap<String, Object>() {{
            put("a", "b");
        }});
```

**Pull request merged**